### PR TITLE
Fix a NullReferenceException in DataFlowAnalysis.VariablesDeclared

### LINKED ISSUE #17281: NullReferenceException when selecting a switch label with errors in pattern matching
**Version Used**: 
VS 2017 d15rel 26214.1
**Steps to Reproduce**:

1. paste the following code in a C# file
```csharp
class A { }

class Test
{
    private void Repro(A node)
    {
        switch (node)
        {
            case A _:
                break;
            case Unknown:
                break;
            default:
                return;
        }
    }
}
```
2. double click to select "Unknown"

**Expected Behavior**:
No error
**Actual Behavior**:
Error about extract method code refactoring provider

```
System.NullReferenceException : Object reference not set to an instance of an object.
   at Microsoft.CodeAnalysis.CSharp.VariablesDeclaredWalker.NoteDeclaredPatternVariables(BoundPattern pattern)
   at Microsoft.CodeAnalysis.CSharp.VariablesDeclaredWalker.VisitPattern(BoundExpression expression,BoundPattern pattern)
   at Microsoft.CodeAnalysis.CSharp.PreciseAbstractFlowPass`1.VisitPatternSwitchBlock(BoundPatternSwitchStatement node)
   at Microsoft.CodeAnalysis.CSharp.PreciseAbstractFlowPass`1.VisitPatternSwitchStatement(BoundPatternSwitchStatement node)
   at Microsoft.CodeAnalysis.CSharp.ControlFlowPass.VisitPatternSwitchStatement(BoundPatternSwitchStatement node)
   at Microsoft.CodeAnalysis.CSharp.BoundPatternSwitchStatement.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.CSharp.PreciseAbstractFlowPass`1.VisitWithStackGuard(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.PreciseAbstractFlowPass`1.VisitAlways(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.AbstractRegionControlFlowPass.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.ControlFlowPass.VisitStatement(BoundStatement statement)
   at Microsoft.CodeAnalysis.CSharp.PreciseAbstractFlowPass`1.VisitStatements(ImmutableArray`1 statements)
   at Microsoft.CodeAnalysis.CSharp.PreciseAbstractFlowPass`1.VisitBlock(BoundBlock node)
   at Microsoft.CodeAnalysis.CSharp.BoundBlock.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.CSharp.PreciseAbstractFlowPass`1.VisitWithStackGuard(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.PreciseAbstractFlowPass`1.VisitAlways(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.AbstractRegionControlFlowPass.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.PreciseAbstractFlowPass`1.Scan(Boolean& badRegion)
   at Microsoft.CodeAnalysis.CSharp.ControlFlowPass.Scan(Boolean& badRegion)
   at Microsoft.CodeAnalysis.CSharp.PreciseAbstractFlowPass`1.Analyze(Boolean& badRegion)
   at Microsoft.CodeAnalysis.CSharp.VariablesDeclaredWalker.Analyze(CSharpCompilation compilation,Symbol member,BoundNode node,BoundNode firstInRegion,BoundNode lastInRegion)
   at Microsoft.CodeAnalysis.CSharp.CSharpDataFlowAnalysis.get_VariablesDeclared()
   at Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.Analyzer.GenerateVariableInfoMap(SemanticModel model,DataFlowAnalysis dataFlowAnalysisData,Dictionary`2 symbolMap)
   at async Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.Analyzer.AnalyzeAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.ExtractMethodAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.ExtractMethod.AbstractExtractMethodService`3.ExtractMethodAsync[TValidator,TExtractor,TResult](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.ExtractMethod.ExtractMethodCodeRefactoringProvider.GetCodeActionAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.ExtractMethod.ExtractMethodCodeRefactoringProvider.ComputeRefactoringsAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.CodeRefactoringService.GetRefactoringFromProviderAsync(<Unknown Parameters>)
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
```

### PR DESCRIPTION
**Pull request by gafter**
Fixes #17281 

@dotnet/roslyn-compiler May I please have a couple of reviews of this very tiny crash fix?

### CODE REVIEW

File: src/Compilers/CSharp/Portable/FlowAnalysis/VariablesDeclaredWalker.cs [hunk 1]
```diff
@@ -77,7 +77,8 @@ private void NoteDeclaredPatternVariables(BoundPattern pattern)
             if (IsInside && pattern.Kind == BoundKind.DeclarationPattern)
             {
                 var decl = (BoundDeclarationPattern)pattern;
-                if (decl.Variable.Kind == SymbolKind.Local)
+                // The variable may be null if it is a discard designation `_`.
+                if (decl.Variable?.Kind == SymbolKind.Local)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/VariablesDeclaredWalker.cs]: >Variable [](start = 25, length = 8)

Are all other references to this property throughout the compiler safe?

Comment (gafter) [inline src/Compilers/CSharp/Portable/FlowAnalysis/VariablesDeclaredWalker.cs]: I was not able to find any other instances of this issue.

**Comment by gafter** [comment]
@dotnet-bot Please test windows_debug_unit32_prtest

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a couple of reviews of this very tiny crash fix?

**Comment by cston** [comment]
LGTM

**Pull request merged**